pkg/controller/plan/migration.go

```diff
@@ -706,11 +709,34 @@ func (r *Migration) execute(vm *plan.VMStatus) (err error) {
 
 		switch vm.Phase {
 		case CreateInitialSnapshot:
-			vm.Phase = CreateDataVolumes
+			vm.Phase = WaitForInitialSnapshot
 		case CreateSnapshot:
-			vm.Phase = AddCheckpoint
+			vm.Phase = WaitForSnapshot
 		case CreateFinalSnapshot:
-			vm.Phase = AddFinalCheckpoint
+			vm.Phase = WaitForFinalSnapshot
+		}
+	case WaitForInitialSnapshot, WaitForSnapshot, WaitForFinalSnapshot:
+		step, found := vm.FindStep(r.step(vm))
+		if !found {
+			vm.AddError(fmt.Sprintf("Step '%s' not found", r.step(vm)))
+			break
+		}
+		snapshot := vm.Warm.Precopies[len(vm.Warm.Precopies)-1].Snapshot
+		ready, err := r.provider.CheckSnapshotReady(vm.Ref, snapshot)
+		if err != nil {
+			step.AddError(err.Error())
+			err = nil
+			break
+		}
+		if ready {
+			switch vm.Phase {
```

Comment:
I think this can be replaced with `vm.Phase = r.next(vm.Phase)` as well if the `WaitFor...` phases are added to the right spots in the warm itinerary.